Test(web,web-react,web-twig): Introduce visual tests of all components #DS-660 (#1196)
lmc-eu/spirit-design-system: CONTRIBUTING.md

**Prerequisites:** [Docker][docker] 🐳

You can run visual regression testing via `Makefile` in the project root.

> All commands will execute docker command that starts [Playwright][playwright] in containerized environment.

Reviewer: ```suggestion
All commands will execute Docker command that starts [Playwright][playwright] in containerized environment.
```

- Using `make test-e2e` you will run the tests.
- Using `make test-e2e-update` you will update the snapshots.

> - Using `make test-e2e-report` you will generate and serve report of the visual regression testing.

Reviewer: Maybe just "Use XY to run/update/generate …"?

> - Using `make test-e2e-update` you will update the snapshots.
> - Using `make test-e2e-report` you will generate and serve report of the visual regression testing.

Reviewer: I think there is no article here:

```suggestion
- Using `make test-e2e-update` you will update snapshots.
- Using `make test-e2e-report` you will generate and serve report of visual regression testing.
```

> - Using `make test-e2e-report` you will generate and serve report of the visual regression testing.

Reviewer: Where is the report served? Is it a URL that is automatically opened in a browser?

Reviewer 2: Yes, the Playwright gives you the URL in the console.

👉 Visual snapshots are generated based on platform, so we need to use same platform locally and on CI (GitHub Actions).

⚠️ Version number of the [Playwright][playwright] dependency must be the same in `package.json` file and in the `./bin/make/e2e.sh` to ensure that no additional [Playwright][playwright] dependencies will need to install (browsers are backed in the Docker image). See https://playwright.dev/docs/docker.

> We run visual regression testing locally against our demo apps. Web and Web React packages are served using Vite.

Reviewer: (I'm commenting on the unchanged paragraph): I think these occurrences of "Playwright" don't need to be links.  It's OK to have the link just in the first occurrence on the page.

Web Twig package is served using Symfony app.

In CI we use Netlify to test against.

> ⚠️ Currently we do not deploy the Web Twig package to any environment, so you can only test it locally.

Reviewer: 👍🏻

> We have two tests and you can find them in the `./tests/e2e` directory:

Reviewer: Isn't it bigger than just a "test"? Something like a test suite maybe?

> - `demo-homepages` - tests homepages of our demo apps'

Reviewer: ```suggestion
- `demo-homepages` - tests the homepages of our demo apps.
```

or

```suggestion
- `demo-homepages` - tests our demo apps' homepages.
```

>   - This test is used to verify that the demo apps are working properly and that their homepages are not broken

Reviewer: ```suggestion
  - This test is used to verify that the demo apps are working properly and their homepages are not broken.
```

> - `demo-components-compare` - tests components' pages of our demo apps'

Reviewer: ```suggestion
- `demo-components-compare` - tests components' pages of our demo apps.
```

>   - This test gets the list of components from file system for each package and then it goes through each component and compares it's page in all demo apps
>   - Only one screenshot is taken for each component, if you run update command, only the last screenshot will be saved

Reviewer: ```suggestion
  - This test gets the list of components from file system for each package and then it goes through each component and compares its page in all demo apps.
  - Only one screenshot is taken for each component. If you run the update command, only the last screenshot will be saved.
```

> 👉 To save time and repository size, we test only in Chromium browser and only on desktop viewport.

Reviewer: 👍🏻
